packages/account-tree-controller/src/backup-and-sync/service/atomic-sync-queue.ts

```diff
@@ -0,0 +1,123 @@
+import { backupAndSyncLogger } from '../../logger';
+import type { AtomicSyncEvent } from '../types';
+
+/**
+ * Manages atomic sync operations in a queue to prevent concurrent execution
+ * and ensure proper ordering of sync events.
+ */
+export class AtomicSyncQueue {
+  /**
+   * Queue for atomic sync events that need to be processed asynchronously.
+   */
+  readonly #queue: AtomicSyncEvent[] = [];
+
+  /**
+   * Flag to prevent multiple queue processing operations from running concurrently.
+   */
+  #isProcessingInProgress = false;
+
+  /**
+   * Clears the queue and enqueues a new sync function.
+   *
+   * @param syncFunction - The sync function to enqueue.
+   * @returns A Promise that resolves when the sync function completes.
+   */
+  clearAndEnqueue(syncFunction: () => Promise<void>): Promise<void> {
+    this.clear();
+    return this.enqueue(syncFunction);
+  }
+
+  /**
+   * Enqueues an atomic sync function for processing.
+   *
+   * @param syncFunction - The sync function to enqueue.
+   * @returns A Promise that resolves when the sync function completes.
+   */
+  enqueue(syncFunction: () => Promise<void>): Promise<void> {
+    let resolvePromise: (() => void) | undefined;
+    let rejectPromise: ((error: unknown) => void) | undefined;
+
+    // Create promise that resolves when the sync function completes
+    const promise = new Promise<void>((resolve, reject) => {
+      resolvePromise = resolve;
+      rejectPromise = reject;
+    });
```

Comment:
Nit: We could have used https://github.com/MetaMask/utils/blob/main/src/promise.ts#L43